api/src/main/java/org/apache/iceberg/expressions/Evaluator.java

```diff
@@ -134,18 +134,18 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
     }
 
     @Override
-    public <T> Boolean in(BoundReference<T> ref, Literal<T> lit) {
-      throw new UnsupportedOperationException("In is not supported yet");
+    public <T> Boolean in(BoundReference<T> ref, LiteralSet<T> literalSet) {
+      return literalSet.contains(ref.get(struct));
     }
 
     @Override
-    public <T> Boolean notIn(BoundReference<T> ref, Literal<T> lit) {
-      return !in(ref, lit);
+    public <T> Boolean notIn(BoundReference<T> ref, LiteralSet<T> literalSet) {
+      return !in(ref, literalSet);
     }
 
     @Override
     public <T> Boolean startsWith(BoundReference<T> ref, Literal<T> lit) {
-      return ((String) ref.get(struct)).startsWith((String) lit.value());
+      return ((String) ref.get(struct)).startsWith(lit.value().toString());
```

Comment:
This isn't a related change, and the literal is guaranteed to be a string at this point. I don't think this should be included in this PR.

Reply:
The issue is that `lit.value()` in this case will return a `CharSeqWrapper` instead of `String`. It cannot be cast to `String`.
If the underlying is a `String`, `toString()` will still return it without extra costs.

Reply:
Why did the type used by the literal change? I don't think it needs to.

Reply:
Here, I changed the construction of `StringLiteral` to be
```    StringLiteral(CharSequence value) {
      super(new CharSeqWrapper(value));
    }```

The other way is to only wrap `CharSequence` within LiteralSet. In that way, there is no need to change `startsWith`.

This seems cleaner with less side effect and I will update the PR accordingly.